near-sdk/src/collections/tree_map.rs

```diff
@@ -0,0 +1,1681 @@
+        self.val.get(key).is_some()
+    }
+
+    pub fn get(&self, key: &K) -> Option<V> {
+        self.val.get(key)
+    }
+
+    pub fn insert(&mut self, key: &K, val: &V) -> Option<V> {
+        if self.contains_key(&key) {
+            // key is already present, changing only associated value
+            self.val.insert(&key, &val)
+        } else {
+            let at = self.root;
+            let id = self.len;
+            let root = self.insert_at(at, id, &key);
+            self.set_root(root);
+            self.len += 1;
+            self.val.insert(&key, &val)
+        }
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        if self.contains_key(&key) {
+            let root = self.do_remove(&key);
+            self.set_root(root);
+            self.len -= 1;
+            self.val.remove(&key)
+        } else {
+            // no such key, nothing to do
+            None
+        }
+    }
+
+    pub fn min(&self) -> Option<K> {
+        self.min_at(self.root, self.root).map(|(k, _, _)| k)
+    }
+
+    pub fn max(&self) -> Option<K> {
+        self.max_at(self.root, self.root).map(|(k, _, _)| k)
+    }
+
+    pub fn floor(&self, key: &K) -> Option<K> {
+        self.floor_at(self.root, key)
+    }
+
+    pub fn ceil(&self, key: &K) -> Option<K> {
+        self.ceil_at(self.root, key)
+    }
+
+    pub fn iter<'a>(&'a self) -> impl Iterator<Item = (K, V)> + 'a {
+        Cursor::asc(&self).into_iter()
+    }
+
+    pub fn iter_from<'a>(&'a self, key: K) -> impl Iterator<Item = (K, V)> + 'a {
+        Cursor::asc_from(&self, key).into_iter()
+    }
+
+    pub fn iter_rev<'a>(&'a self) -> impl Iterator<Item = (K, V)> + 'a {
+        Cursor::desc(&self).into_iter()
+    }
+
+    pub fn iter_rev_from<'a>(&'a self, key: K) -> impl Iterator<Item = (K, V)> + 'a {
+        Cursor::desc_from(&self, key).into_iter()
+    }
+
+    /// Iterate over K keys in ascending order in O(Klog(N))
+    ///
+    /// # Panics
+    ///
+    /// Panics if range start > end.
+    /// Panics if range start == end and both bounds are Excluded.
+    pub fn range<'a>(&'a self, r: (Bound<K>, Bound<K>)) -> impl Iterator<Item = (K, V)> + 'a {
+        let (lo, hi) = match r {
+            (Bound::Included(a), Bound::Included(b)) if a >  b => panic!("Invalid range."),
+            (Bound::Excluded(a), Bound::Included(b)) if a >  b => panic!("Invalid range."),
+            (Bound::Included(a), Bound::Excluded(b)) if a >  b => panic!("Invalid range."),
+            (Bound::Excluded(a), Bound::Excluded(b)) if a == b => panic!("Invalid range."),
+            (lo, hi) => (lo, hi)
+        };
+
+        Cursor::range(&self, lo, hi).into_iter()
+    }
+
+    pub fn to_vec(&self) -> Vec<(K, V)> {
+        self.iter().collect()
+    }
+
+    //
+    // Internal utilities
+    //
+
+    fn set_root(&mut self, root: u64) {
+        env::storage_write(&self.tree_prefix, &serialize(&root));
+        self.root = root;
```

Comment:
we shouldn't have to `storage_write`: `Vector` doesn't do it with `len` field. CC @nearmax

Reply:
`root` is rather important property of a tree, not saving it would require special kind of lookup (most likely in O(N)) to restore it for an existing tree.

Reply:
It's a bit confusing:
- simple classes don't have to handle their own storage they are stored by the owner (some key determined by the owner -> serialized object)
- near_sdk collections manage their storage and their serialization doesn't include elements
- they still have fields like `len` in Vector, and rely on those fields to be stored/read by the owner serializing the collection: `Vector::new(id)` sets `len=0`, so it can only be used for creating a new vector as opposed to providing access to an existing collection prefixed by `id`
- `TreeMap::new` calls `UnorderedMap::new` which in turn calls `Vector::new`. So we can't use `TreeMap::new` to access an existing map: even with reading `root` the `len` field is broken.

Reply:
makes sense, removing store/load of root id

Reply:
done